core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala

```diff
@@ -29,6 +29,38 @@ import org.junit.Test
 
 class ConsoleConsumerTest {
 
+  @Test
+  def shouldResetOffsetsBeforeExitForNewConsumer() {
+    //Given
+    val args: Array[String] = Array(
+      "--bootstrap-server", "localhost:9092",
+      "--topic", "test",
+      "--group", "test-group",
+      "--max-messages", "123")
+
+    //When
+    val config = new ConsoleConsumer.ConsumerConfig(args)
+
+    // Mocks
+    val consumer = EasyMock.createNiceMock(classOf[NewShinyConsumer])
+
+    // Stubs
+    val record = new BaseConsumerRecord(topic = "foo", partition = 0, offset = 1, key = Array[Byte](), value = Array[Byte]())
+
+    // Expectations
+    ConsoleConsumer.messageCount = 0 // reset message count to zero
+    EasyMock.expect(consumer.receive()).andReturn(record).times(config.maxMessages)
```

Comment:
I think it's going to be difficult triggering the new logic through a mock of `NewShinyConsumer`. We might be able to do it using `MockConsumer` if we change `NewShinyConsumer` to take an instance of `Consumer` through the constructor rather than constructing `KafkaConsumer` internally.